tests/Functional/Driver/Mysqli/ResultTest.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\DBAL\Tests\Functional\Driver\Mysqli;
+
+use Doctrine\DBAL\Driver\Mysqli\Result;
+use Doctrine\DBAL\Schema\Table;
+use Doctrine\DBAL\Tests\FunctionalTestCase;
+use Doctrine\DBAL\Tests\TestUtil;
+use mysqli;
+use mysqli_driver;
+use mysqli_sql_exception;
+
+use function assert;
+use function sprintf;
+
+use const MYSQLI_REPORT_ERROR;
+use const MYSQLI_REPORT_OFF;
+use const MYSQLI_REPORT_STRICT;
+
+/** @requires extension mysqli */
+final class ResultTest extends FunctionalTestCase
+{
+    private const TABLE_NAME = 'result_test_table';
+
+    private mysqli $nativeConnection;
+
+    protected function setUp(): void
+    {
+        if (! TestUtil::isDriverOneOf('mysqli')) {
+            self::markTestSkipped('This test requires the mysqli driver.');
+
+            return;
+        }
+
+        $nativeConnection = $this->connection->getNativeConnection();
+
+        assert($nativeConnection instanceof mysqli);
+
+        $this->nativeConnection = $nativeConnection;
+
+        $table = new Table(self::TABLE_NAME);
+        $table->addColumn('my_col_1', 'integer', ['notnull' => true]);
+
+        $this->dropAndCreateTable($table);
+    }
+
+    protected function tearDown(): void
+    {
+        $this->dropTableIfExists(self::TABLE_NAME);
+    }
+
+    public function testSuccessfulRowCountFromAffectedRows(): void
+    {
+        $mysqliStmt = $this->nativeConnection
+            ->prepare(sprintf('INSERT INTO %s VALUES (7);', self::TABLE_NAME));
+        $mysqliStmt->execute();
+
+        self::assertSame(1, $mysqliStmt->affected_rows);
+        self::assertSame(1, (new Result($mysqliStmt))->rowCount());
+    }
+
+    public function testFailingRowCountFromAffectedRows(): void
+    {
+        $mysqliStmt = $this->nativeConnection
+            ->prepare(sprintf('INSERT INTO %s VALUES (NULL);', self::TABLE_NAME));
+
+        $mysqliDriver = new mysqli_driver();
+
+        // Ensure default configuration present in PHP < 8.1.
+        $mysqliDriver->report_mode = MYSQLI_REPORT_OFF;
+
+        $mysqliStmt->execute();
+
+        self::assertSame(-1, $mysqliStmt->affected_rows);
+        self::assertSame(-1, (new Result($mysqliStmt))->rowCount());
+
+        // Ensure default configuration present in PHP >= 8.1.
+        // @see: https://php.watch/versions/8.1/mysqli-error-mode.
+        $mysqliDriver->report_mode = MYSQLI_REPORT_ERROR | MYSQLI_REPORT_STRICT;
+
+        $error = null;
```

Comment:
Anything beyond this point looks pointless to me. We should make sure we're still testing DBAL and not the MySQLi extension itself.

Reply:
IMO, this is the key point in this test and the covered behavior. Currently we are letting the MySQLi behavior to bubble up to the DBAL layer in the `Result` object. Therefore, we need to make a test that covers DBAL, even though it seems to cover MySQLi..
What I'm trying to do in #5915 is to mask the MySQLi behavior to provide a consistent API from DBAL, using exceptions instead of returning "magic" values when errors occur.